File: transliteration.c
#define _CRT_SECURE_NO_WARNINGS
#define MAX_LEN_WORD 25
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

const char* alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
const char* numbers = "0123456789";

// Классификация символа
const char* classifySymbol(char symbol)
{
    switch (symbol)
    {
    case '$': return "доллар";
    case ';': return "тчкзпт";
    case '=': return "равно";
    case '\'': return "одинкавыч";
    case '-': return "тире";
    case ' ': return "пробел";

    default: return "ошибка";
    }
}

// Транслитерация символьной цепочки в цепочку лексем (подается строка)
char** trl(char* str)
{
    int length = strlen(str);
    char** arrSymbols = malloc(length * sizeof(char*));

    const char sym[] = { '$', ';', '=', '\'', '-', ' ' };
    const char* symbolClass = NULL;

    for (int i = 0; i < length; i++)
    {
        char symbol = str[i];

        if (strchr(alphabet, symbol))
            symbolClass = "буква";
        else if (strchr(numbers, symbol))
            symbolClass = "цифра";
        else
        {
            int check = 0;
            for (int j = 0; j < sizeof(sym) / sizeof(sym[0]); j++)
            {
                if (sym[j] == symbol)
                {
                    symbolClass = classifySymbol(symbol);
                    check = 1;
                    break;
                }
            }
            if (check == 0)
                symbolClass = "ошибка";
        }

        arrSymbols[i] = malloc(2 * sizeof(char));
        arrSymbols[i][0] = symbol;
        strcpy(arrSymbols[i] + 1, symbolClass);
    }

    return arrSymbols;
}

File: transliteration.py
def classify_symbol(symbol): #Классификация спецсимволов
    if symbol == '$':
        return "доллар"
    elif symbol == ';':
        return "тчкзпт"
    elif symbol == '=':
        return "равно"
    elif symbol == '\'':
        return "одинкавыч"
    elif symbol == ' ':
        return "пробел"
    elif symbol == '.':
        return "точка"
    elif symbol == '-':
        return "минус"
    elif symbol == "+":
        return "плюс"
def tansliteration(string):
     alphabet = list("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz") #Список алфавита
     numbers = list("0123456789") #Список цифр
     sym = ['$', ';', '=', '\'', ' ', '-', '+', '.'] #Список спецсимволов

     arrSymbols = [] #Список для хранения символов с их классами
     i = 0
     while i < len(string):
         if string[i] in alphabet:
             alp = (string[i], "буква")
             arrSymbols.append(alp)
         elif string[i] in numbers:
             num = (string[i], "цифра")
             arrSymbols.append(num)
         elif string[i] in sym:
             spec = (string[i], classify_symbol(string[i]))
             arrSymbols.append(spec)
         else:
             err = (string[i], "ошибка")
             arrSymbols.append(err)
             print(f"Недопустимый символ: {string[i]}")
         i += 1

     return arrSymbols
